presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java

```diff
@@ -41,12 +44,48 @@
     private final AtomicReference<PrestoConnection> connection;
     private final AtomicReference<ResultSet> currentResult = new AtomicReference<>();
     private final AtomicLong currentUpdateCount = new AtomicLong(-1);
+    private final ProgressCallback progressCallback = new ProgressCallback();
+
+    @ThreadSafe
+    final class ProgressCallback
+    {
+        private final AtomicReference<Consumer<QueryStats>> callback = new AtomicReference<>();
+
+        public void setCallback(Consumer<QueryStats> progressMonitor)
+        {
+            this.callback.set(progressMonitor);
+        }
+
+        public void clearCallback()
+        {
+            this.callback.set(null);
+        }
+
+        void issueCallback(QueryStats queryStats)
+        {
+            Consumer<QueryStats> callback = this.callback.get();
+            if (callback == null) {
+                return;
+            }
+            callback.accept(queryStats);
+        }
+    }
 
     PrestoStatement(PrestoConnection connection)
     {
         this.connection = new AtomicReference<>(requireNonNull(connection, "connection is null"));
     }
 
+    public void setProgressMonitor(Consumer<QueryStats> progressMonitor)
+    {
+        this.progressCallback.setCallback(progressMonitor);
```

Comment:
Check for null (technically passing null works fine, but that isn't valid use of the API)